### PR TITLE
SNOW-698916 support Python 3.11

### LINKED ISSUE #1345: SNOW-698916: Pyarrow 8 is not compatible with Python 3.11
Pyarrow 10.0.1 just released today (supports Python 3.11), but snowflake-connector-python is bound to `pyarrow>=8.0.0,<8.1.0` which makes it impossible to run on Python 3.11.

1. What version of Python are you using?

   `Python 3.11.0 (main, Oct 24 2022, 22:24:46) [Clang 14.0.0 (clang-1400.0.29.102)]`

2. What operating system and processor architecture are you using?

   `macOS-13.0.1-arm64-arm-64bit`

3. What are the component versions in the environment (`pip freeze`)?

   `pyarrow==10.0.1`

4. What did you do?

   Install `snowflake-connector-python`

### PR DESCRIPTION
**Pull request by noamcohen97**
Fixes #1345

@sfc-gh-aling

### CODE REVIEW

**Comment by noamcohen97** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by noamcohen97** [comment]
recheck

File: setup.py
Comment (3dbrows) [inline setup.py:75]: We also need: `libarrow_dataset.so.1000`

File: setup.py
Comment (3dbrows) [inline setup.py:80]: We also need: `libarrow_dataset.1000.dylib`

File: setup.py
Comment (3dbrows) [inline setup.py:84]: We also need: `arrow_dataset.dll`

File: setup.py
Comment (3dbrows) [inline setup.py:89]: We also need: libarrow_dataset.so.1000

File: setup.py
Comment (3dbrows) [inline setup.py:94]: We also need: `libarrow_dataset.1000.dylib`

File: setup.py
Comment (3dbrows) [inline setup.py:98]: We also need: `arrow_dataset.lib`

Comment (3dbrows) [inline setup.py:98]:
```suggestion
            "win32": [
                "arrow.lib",
                "arrow_dataset.lib",
                "arrow_python.lib", 
                "parquet.lib"
            ],
```

**Review by 3dbrows — CHANGES_REQUESTED** [review]
Added some commit suggestions for the `libarrow_dataset` library which appears to be necessary with Arrow 10.

**Comment by potiuk** [comment]
:crossed_fingers:

**Comment by idanmiara** [comment]
Thanks @noamcohen97 
I can confirm that it builds and installs locally on Python 3.11 on macos.

**Comment by noamcohen97** [comment]
@sfc-gh-mkeller Can you please have a look?

File: setup.py
Comment (sfc-gh-mkeller) [inline setup.py:152]: @xhochy added this originally, I'd like him to take a look too

**Review by sfc-gh-mkeller — COMMENTED** [review]
Note for the team:
* We'll need to update manylinux2014_x86_64 docker images on our internal mirror ✅ 
* We need to setup Python 3.11 on Mac and Windows Jenkins workers and seed the new 3.11 jobs
* Generate new Jenkins test jobs
* Probably update `gosu` versions in the Dockerfiles ✅ 

Otherwise I think the PR itself looks good!
@noamcohen97 would you mind bumping the `gosu` version to the newest one in your PR too?

File: setup.py
Comment (noamcohen97) [reply setup.py:152]: here are the documentation pages stating the requirement for for C++17
https://arrow.apache.org/docs/developers/cpp/building.html#system-setup
https://arrow.apache.org/docs/cpp/conventions.html#language-version

**Comment by noamcohen97** [comment]
Thanks for reviewing @sfc-gh-mkeller!


> would you mind bumping the `gosu` version to the newest one in your PR too

Sure! Will do.


Also, would you mind approving the workflows so build and tests will run?

**Comment by noamcohen97** [comment]
BTW, would you like me to add an entry to the release notes?

File: setup.py
Comment (xhochy) [reply setup.py:152]: I mostly added the `if` to not override the `std` if it is passed in via `CXXFLAGS`. As we're keeping that, I'm happy. The change of `std` for the new Arrow is also correct.

**Comment by noamcohen97** [comment]
@sfc-gh-mkeller any updates on this?

**Comment by dataders** [comment]
hi @sfc-gh-aling and @sfc-gh-mkeller sorry to "at" you here, but we're blocked on releasing the next minor version of dbt-core until this is PR is merged and released. :eyes: @colin-rogers-dbt

**Comment by sfc-gh-mkeller** [comment]
> BTW, would you like me to add an entry to the release notes?

Yes, absolutely!

> @sfc-gh-mkeller any updates on this?

We're currently blocked on our Windows and Mac Jenkins workers having Python 3.11 installed. We have completed the rest of the steps necessary to merge this. That being said our target to release Python 3.11 support is in the January release.

**Comment by noamcohen97** [comment]
> our target to release Python 3.11 support is in the January release.

@sfc-gh-mkeller Sounds great! looking forward for it

**Comment by sfc-gh-mkeller** [comment]
`pyarrow==10.0.1` support was merged through #1347
This PR will need to be rebased

**Comment by noamcohen97** [comment]
> > BTW, would you like me to add an entry to the release notes?
> 
> Yes, absolutely!

Done!

**Comment by Yarden-zamir** [comment]
Hi, whatsup with this? I really would love to have 3.11

**Comment by yogevyuval** [comment]
@sfc-gh-mkeller Hi! Any news regarding this? We would love to test python 3.11

**Comment by Yarden-zamir** [comment]
Note for people who need this but are not patient, you can use the branch of this pr in your requirements / pip install like so
```
git+https://github.com/qlik-trial/qlik-qdi-sdk-python.git
```

**Review by sfc-gh-mkeller — APPROVED** [review]
:ship: :ship:

**Pull request merged**
